Fixes for General Additive Models

rogancarr opened the pull request:
This PR addresses a number of issues with the `General Additive Model` (`GAM`) trainer. In particular, it addresses the issues with the `GAM Classifier` not fitting nor producing a probability, and adds support for validation pruning, summary text, and centering of the feature effects around a mean response (e.g. intercept).

Additionally, the PR addresses some minor issues in the codebase, like `GAMs` using copy-and-paste versions of `FastTree` code, unnecessarily `public` attributes, unused arguments, and splits the `BinaryClassifier` and `Regressor` into separate files.

The changes are as follows:
1. Centered the response and added an intercept term; fixed corresponding issues with table lookups, sparse calculations, and exports. (#739)
2. Added in a validation set and pruning based on the validation set, producing a final pruned graph at the end. (#740)
  a. Switched to using `ScoreTrackers` to keep track of scores during boosted
  b. Save boost iterations as individual graphs (n_features x n_iterations x n_boosts) for pruning.
3. Fixed `GAM Classifier` to use a small learning rate (#741)
  Updated the `FastTreeBinaryClassification` logistic loss gradient to take the `sigmoid` parameter as input: `GAM Binary Classifier` now uses unity; `FastTree Binary Classifier` uses the same default of `2*learning rate` (no change). Optionally, we can plumb this to the FastTree arguments if we show an experimental gain; We can also experiment to see if the `sigmoid` parameter gives gains for GAMs (i.e. by slowing learning even more).
7. Added calibration to the `GamClassifier` to produce probabilities (#738).
  This meant changing the various interfaces.
8. Refactored `GAMRegressor` and `GAMClassifier` into their own files.
  The one file for all GAM trainers had gotten a bit long, and this change is consistent with ML.NET tradition.
9. Added tests to verify train loss and validation metrics (e.g. same value is produced by boosting and the lookup table).
10. Added a Summary to the `GAMPredictor` to produce training statistics and the feature table (#742).
11. Rewrote the core split finding routine to use the same code as FastTree to rely on FastTree unit tests
  The `GAM` routines used a copy-and-paste of internal `FastTree` components. To fight entropy, these were refactored to use the same central calculation, such that it is verified and validated by the `FastTree` unit and end-to-end tests.
12. Removed unused arguments.
13. Allowed for weighted samples.

Fixes #738 
Fixes #739 
Fixes #740 
Fixes #741 
Fixes #742

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=743) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=743) before we can accept your contribution.<br/><br/>:x: rogancarr [sign now](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=743)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/machinelearning?pullRequest=743) it.</sub>

TomFinley commented on src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj:
```diff
@@ -27,6 +32,8 @@
     <Compile Include="FastTreeTweedie.cs" />
+    <Compile Include="GamClassification.cs" />
+    <Compile Include="GamRegression.cs" />
     <Compile Include="GamTrainer.cs" />
```
Actually this is somewhat peculiar. I know you didn't do this, but why is the explicit listing in this project necessary at all? Does anything break if you get rid of this entire `ItemGroup` with all the `<Compile Includes`?

TomFinley commented:
Thanks for doing this @rogancarr -- should I infer from the lack of baseline changes that no tests were migrated to test GAM? I think we probably ought to do some here -- we have some standard regression problems and whatnot.

Ivanidzo4ka commented on test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs:
```diff
@@ -0,0 +1,305 @@
+        {
+            using (var env = new TlcEnvironment())
+            {
+                var trainFile = "binary_sine_logistic_10k.tsv";
```
>binary_sine_logistic_10k [](start = 33, length = 24)

Just a reminder what you need to check in this file into repo, or use build actions to download it.
(if you want to check it in, you also need CELA approval)

TomFinley replied:
Hi @Ivanidzo4ka , I believe that this is a synthetic dataset, so probably CELA is not necessary, since it does not come from any real external source? Or am I wrong?

rogancarr replied:
This is a synthetic dataset. If we want to add this test to the repo, then I'll add the file.

rogancarr replied:
But more important question is: Do we need this test?

This is really a test on the internal state of the learner that I used for debugging. If we add baseline tests for `GAMs`, then it seems like we don't need this test, and we may not want it, since it relies on the internal state.

TomFinley commented:
> 	<EnableDefaultCompileItems>False</EnableDefaultCompileItems>

I think this is the culprit here, that is requiring the explicit listing.

---
Refers to: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj:7 in d0d4e6b. [](commit_id = d0d4e6b76eacd1873d3d463635c18aa7277603f1, deletion_comment = False)

rogancarr commented:
Yes, my best guess is that this line:
```<EnableDefaultCompileItems>False</EnableDefaultCompileItems>```
is why we are explicitly listing the files to compile.

rogancarr commented:
New commit addressing offline reviews:
- Switched sub-graph calculation from multidimensional arrays to structs
- Removed TrainingInfo (and therefor the test to validate training loss): We need a general way to do this and not a one-off solution for one learner.
- Removed post-training scoring update: At this point it's not necessary and hurts performance
- Made the statistics calculator interface internal to FastTree: It cannot be an abstract class

Next:
- Running performance comparisons against the previous version for GAM{Regressor, Binary Classifier} and FastTree{Regressor, and Binary Classifer}
- Recomputing baselines

TomFinley commented on src/Microsoft.ML.FastTree/Dataset/FeatureFlock.cs:
```diff
@@ -248,7 +248,7 @@ public void CopyFeatureHistogram(int subfeatureIndex, ref PerBinStats[] hist)
             int gtCount = 0;
-            sumWeights = 2 * eps;
+            sumWeights += 2 * eps;
             double gainShift = leafCalculator.GetLeafSplitGain(totalCount, sumTargets, sumWeights);
```
Nice

rogancarr commented:
New commits
- Fixed an introduced bug in the split finding algorithm for weighted calculations
- Updated the baseline tests for GamBinaryClassifier

I performed two performance tests, and sent results to reviewers offline:
1. Determine if the new implementation of GAMs had an effect on goodness-of-fit (AUC, RMS) and training time
2. Determine if the changes to FastTree had an effect on goodness-of-fit (AUC, RMS) and training time.

Test Results:
1. GAM changes resulted better goodness of fit for Binary Classification, and equivalent for Regression. Regression was not numerically the same because the order of additions to the final graphs changed. The changes also introduced a slowdown on the order of 1.3 due to the way that graphs are held for pruning (many small pieces instead of one master graph). This can be targeted for future optimization (I have a few ideas).
2. The changes had no measurable effect on FastTree. Goodness-of-fit was numerically identical and training time was equivalent.

rogancarr commented:
In the build, [SDCA failed a test on Linux Debug](https://dnceng.visualstudio.com/public/_build/results?buildId=20807&view=ms.vss-test-web.test-result-details) -- this seems unrelated to the changes in this PR -- is it possible to rerun the tests?

Zruty0 commented on src/Microsoft.ML.FastTree/GamRegression.cs:
```diff
@@ -0,0 +1,110 @@
+        {
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Metric for pruning. (For regression, 1: L1, 2:L2; default L2)", ShortName = "pmetric")]
+            [TGUI(Description = "Metric for pruning. (For regression, 1: L1, 2:L2; default L2")]
+            public int PruningMetrics = 2;
```
>int PruningMetrics = 2 [](start = 19, length = 22)

wow, really? Can we maybe have enum here?

rogancarr replied:
This is a holdover from `FastTree` -- this is copied over from `FastTreeRegression`.  It's a parameter to the [`RegressionTest` class](https://github.com/dotnet/machinelearning/blob/a8cd341f8c03985aaea216e8c911c466ad020b1e/src/Microsoft.ML.FastTree/Training/Test.cs#L516).

If possible, I'd like to file this as a separate issue and fix as a separate PR.

Zruty0 commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -209,25 +96,41 @@ public abstract class ArgumentsBase : LearnerInputBaseWithWeight
+        // Training Datastructures
+        private SubGraph _subGraph;
 
         //Results of Training
```
>//Results of Training [](start = 8, length = 21)

please use the common syntax for comments.
//<space>Sentence starting from capital letter (no other capitalized words inside)<period>

rogancarr replied:
Done for all comments!

Zruty0 commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
 
-        public sealed override TPredictor Train(TrainContext context)
+        /// <summary>
+        /// Load the data off the context and train the model
```
>Load the data off the context and train the model [](start = 12, length = 49)

That is not a helpful comment :)

rogancarr replied:
Removed!

Zruty0 commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
+        /// <summary>
+        /// Load the data off the context and train the model
+        /// </summary>
+        /// <param name="context"></param>
```
>param [](start = 13, length = 5)

remove empty param tags

rogancarr replied:
Thanks, -- removed!

Zruty0 commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
 
-        private protected abstract TPredictor CreatePredictor();
+        /// <summary>
+        /// Define the ScoreTrackers to use in the calculation.
```
>Define the ScoreTrackers [](start = 12, length = 24)

these comments are not helpful. I already have a good suspicion that `DefineScoreTrackers` defines score trackers. 

These are private methods too, so might as well not have comments at all

rogancarr replied:
Thanks -- removed!

rogancarr commented:
Thanks for the review @TomFinley  and @Zruty0 ! I addressed all but one issue and responded to one in the comments here. Note that WIP is flagged, but this is not WIP -- I had a WIP commit that I didn't remove because we will squash-merge this branch.
